src/Microsoft.PowerShell.Commands.Management/commands/management/CombinePathCommand.cs

```diff
@@ -119,15 +129,27 @@ protected override void ProcessRecord()
                     continue;
                 }
 
-                if (Resolve)
+                if (Resolve.IsPresent)
                 {
                     // Resolve the paths. The default API (GetResolvedPSPathFromPSPath)
                     // does not allow non-existing paths.
-                    Collection<PathInfo> resolvedPaths = null;
+                    Collection<string> resolvedPaths = null;
```

Comment:
What benefits you expect?
I see extra allocations below.
Also with PathInfo it is easy to enhance the cmdlet in future.

Reply:
It's not so much that reducing this to string has a benefit, it's that the other provider method doesn't give back PathInfo objects to work with.

If there is a way to create PathInfo objects for nonexistent items, I'm all ears.

If not, the other alternative would be to diverge completely here and handle both cases entirely separately, I suppose?

Reply:
I'd don't mess these two code paths - old and new.

Also we have ` public string GetUnresolvedProviderPathFromPSPath(string path)`

Reply:
Fair point, I'll play with the structure. :)

Reply:
@iSazonov that overload doesn't allow us to pass in `CmdletProviderContext`, which means it would not be able to distinguish the different providers, right? Or am I drawing the wrong conclusions there?

Reply:
Oops... If we use SkipValidation under Resolve we must use GetResolvedPSPathFromPSPath() API in any case. So we need enhance the API to do not validate paths.

Reply:
There already exist `allowNonexistingPaths` parameter.

Reply:
I'm not sure I follow. Why do we need to modify the API? At least on the surface it would appear that `GetUnresolvedProviderPathFromPSPath()` performs the same function as `GetResolvedPSPathFromPSPath()`, at least in the context of its use in this cmdlet. After all, `Join-Path` doesn't really use the `PathInfo` object, simply returning the string value from the `PathInfo.Path` property of each resolved path.

The reason such a path must stay _un_resolved (despite the wording of the parameters vs the API) is that if you give it a path segment containing a wildcard, we can't assume that the wildcard refers to only existing items, by the nature of the request.

Reply:
resolved - here means "resolve wildcards".

Reply:
Yes, but if a path doesn't exist a wildcard can't be resolved. Should a different term (different switch entirely?) be used for this kind of operation? It's not really resolving wildcards, but it's validating and resolving relative nodes to a single straightforward path according to the PSProvider's path format

Reply:
> if a path doesn't exist a wildcard can't be resolved.

Can with SkipValidation.   c:\window*\notexisting.txt -> c:\windows\notexisting.txt

Reply:
Yes, assuming that part of the path actually exists. But if we're skipping validation to allow resolution of non-existent paths, we can't assume wildcarded segments will necessarily exist?

E.g., if path `C:\test\` exists with no subfolders and user tries `Join-Path -Path C:\Test ..\Test\*\test2 -Resolve -SkipValidation` should the cmdlet still return `C:\Test\*\Test2`, or should it return nothing as it does currently with -Resolve?

Reply:
I still expect that allowNonexistingPaths is that resolve your questions. Note that current behavior is (1) resolve paths (with wildcards), (2) check existence - we need only turn off second (with allowNonexistingPaths ).

Reply:
And if the path segment that contains the wildcard isn't found? Just return nothing?

I'll look into adding an overload for the provider method there. 🙂

Reply:
@iSazonov OK, looks like it was pretty simple overall...

This is the result:
```powershell
PS Z:\Scripts\PowerShell-Core> "c:\aaa" | Join-Path -ChildPath ".\..\bbb" -Resolve
Join-Path : Cannot find path 'C:\bbb' because it does not exist.
At line:1 char:12
+ "c:\aaa" | Join-Path -ChildPath ".\..\bbb" -Resolve
+            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : ObjectNotFound: (C:\bbb:String) [Join-Path], ItemNotFoundException
+ FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.JoinPathCommand
 
PS Z:\Scripts\PowerShell-Core> "c:\aaa" | Join-Path -ChildPath ".\..\bbb" -Resolve -SkipValidation
C:\bbb
PS Z:\Scripts\PowerShell-Core> "c:\aaa" | Join-Path -ChildPath ".\..\bbb\*" -Resolve -SkipValidation
Join-Path : Cannot find path 'C:\bbb' because it does not exist.
At line:1 char:12
+ "c:\aaa" | Join-Path -ChildPath ".\..\bbb\*" -Resolve -SkipValidation
+            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : ObjectNotFound: (C:\bbb:String) [Join-Path], ItemNotFoundException
+ FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.JoinPathCommand
```

In other words, a wildcard is only permitted in a path segment that currently exists, but otherwise nonexistent relative references are resolved accurately and a properly-formed path results.

I _think_ this is OK? Unsure, to be honest,

Reply:
I think it is right behavior.
Also I'd consider `Join-Path -Path c:\aaa\*\bbb -ChildPath ccc`.

Please contribute and add tests.

Reply:
🤔 OK, added tests and rebased. That last case you mention where the wildcard may refer to multiple existing paths with non-existing leaf node(s)... that's not handled properly. It just returns nothing at the moment. I'll come back to this later on today or perhaps tomorrow. May need to dig deeper into the underlying methods to see how to best handle that one.